[7/N][DTensor device order] Support DCP with shard_order

zpcore opened the pull request:
DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165197
* #165092


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

zpcore committed:
[7/N][DTensor device order] Support DCP with shard_order

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang reviewed (commented):
I'm endorsing the change, it looks plausible (since StridedShard seems to imply DCP already supports reordered sharding), but you need a DCP review

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented on test/distributed/checkpoint/test_state_dict.py:
```diff
@@ -1094,5 +1101,59 @@ def test_no_dist(self) -> None:
+            {1: [1, 0]},  # reversed order on tensor dim 1
+        ],
+    )
+    def test_dtensor_checkpoint_with_shard_order(self, shard_order):
```
Added those two tests, PTAL, thanks!

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented:
I further worked on DCP for AutoParallel in https://github.com/meta-pytorch/autoparallel/pull/206 and noticed that this PR is not sufficient to support DCP. This is because when we work with DCP for torch.nn.module, we will call sth like:
```
model_state_dict, optim_state_dict = get_state_dict(
                    parallel_mod,
                    optimizer,
                )
```
Internally, getting a state dict will trigger aten.detach op to create a dict with reference to DTensor parameters in the model. If our DTensor propagation strategy (especially for `detach`)  [link](https://github.com/pytorch/pytorch/blob/51319ca090bc7458168a8451c04ca7e021a72693/torch/distributed/tensor/_ops/_tensor_ops.py#L59-L69) doesn't propagate the `shard_order` then we will lose the device order information, thus results in incorrect state when resume.

zpcore commented:
Optimizer also doesn't work because optimizer state dict will call op like `toch.zeros_like` as, for example, [here](https://github.com/pytorch/pytorch/blob/eaeaa08e3a8071be46f833f7b46aa642ec14e0f7/torch/optim/adam.py#L178).

So we need to support `shard_order` propagation at least for `detach` (fixes model state dict) and `zeros_like` (fixes optimizer state). I should prioritize on enabling propagating for `shard_order` op strategy first.

zpcore committed:
Update on "[7/N][DTensor device order] Support DCP with shard_order"

DCP for tensor with `shard_order` works as below:
```
+------------------------------------+
|     DTensor (dt) + shard_order     |
+------------------------------------+
         |
         |  calls
         v
+---------------------------------------------+
| torch.distributed.checkpoint.save           |
+---------------------------------------------+
         |
         |  triggers
         v
+-------------------------------------------------------------+
| compute_local_shape_and_global_offset (adjusts shard order) |
+-------------------------------------------------------------+
         |
         |  after save, becomes
         v
+-------------------------+
|      checkpoint file    |
+-------------------------+
         |
         |  (later) call
         v
+---------------------------------------------+
| torch.distributed.checkpoint.load           |
+---------------------------------------------+
         |
         |  on top of
         v
+-------------------+
|  Tensor (empty)   |
+-------------------+
         |
         |  after load, becomes
         v
+----------------------------------------------------------------------+
| Tensor, which equals to full DTensor (dt)  with default device order |
+----------------------------------------------------------------------+
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented:
This PR is obsolete under the new StridedShard plan right?
